File: MiddleCharacter/MiddleChar.js

// Solution

const middleChar = string => {
    const charCount = string.length;
    const isEven = charCount % 2 == 0;
    const oddStartIndex = Math.floor(charCount / 2);
    const endIndex =  oddStartIndex + 1

    return string.slice(oddStartIndex - (isEven ? 1 : 0), endIndex);
}

// Tests

class Tests {

    run() {
        this.testEvenNumberOfChars();
        this.testOddNumberOfChars();
        this.testSingleChar();
    }

    testEvenNumberOfChars() {
        const result = middleChar("test");
        const expected = "es";
        this.compare(result, expected);
    }

    testOddNumberOfChars() {
        const result = middleChar("testing");
        const expected = "t";
        this.compare(result, expected);
    }

    testSingleChar() {
        const result = middleChar("t");
        const expected = "t";
        this.compare(result, expected);
    }

    compare(result, expected) {
        if (result == expected) {
            console.log("Test passed!");
        } else {
            console.log(`Test failed! "${result}" does not equal expected "${expected}"`);
        }
    }

}

new Tests().run();

File: MiddleCharacter/MiddleChar.py

# Solution

def middleChar(string):
    characterCount = len(string)
    oddStartIndex = int(characterCount / 2)
    endIndex = oddStartIndex + 1

    return string[oddStartIndex - (1 if isEven(characterCount) else 0) : endIndex]

def isEven(count):
    return count % 2 == 0

# Tests

class Tests:

    def run(self):
        self.testEvenNumberOfChars()
        self.testOddNumberOfChars()
        self.testSingleChar()

    def testEvenNumberOfChars(self):
        result = middleChar("test")
        expected = "es"
        self.compare(result, expected)

    def testOddNumberOfChars(self):
        result = middleChar("testing")
        expected = "t"
        self.compare(result, expected)
    
    def testSingleChar(self):
        result = middleChar("t")
        expected = "t"
        self.compare(result, expected)

    def compare(self, result, expected):
        if result == expected:
            print("Test passed!")
        else:
            print(f'Test failed! "{result}" does not equal expected: "{expected}"')


Tests().run()